onmt/modules/lora.py

```diff
@@ -106,7 +107,126 @@ def forward(self, x: torch.Tensor):
+                    self.lora_B = nn.Parameter(self.weight.new_zeros((args[1], r)))
+                    self.scaling = self.lora_alpha / self.r
+                    # Freezing the pre-trained weight matrix
+                    self.weight.requires_grad = False
+                self.reset_parameters()
+
+            def reset_parameters(self):
+                super().reset_parameters()
+                if hasattr(self, "lora_A"):
+                    # initialize A the same way as the default
+                    # for nn.Linear and B to zero
+                    nn.init.kaiming_uniform_(self.lora_A, a=math.sqrt(5))
+                    nn.init.zeros_(self.lora_B)
+
+            def train(self, mode: bool = True):
+                super().train(mode)
+                if mode:
+                    if self.merge_weights and self.merged:
+                        # Make sure that the weights are not merged
+                        if self.r > 0:
+                            self.weight.data -= (
+                                self.lora_B @ self.lora_A
+                            ) * self.scaling
+                        self.merged = False
+                else:
+                    if self.merge_weights and not self.merged:
+                        # Merge the weights and mark it
+                        if self.r > 0:
+                            self.weight.data += (
+                                self.lora_B @ self.lora_A
+                            ) * self.scaling
+                        self.merged = True
+
+            def forward(self, x: torch.Tensor):
+                result = super().forward(x)
+                if self.r > 0 and not self.merged:
+                    if self.r > 0:
+                        result += (
+                            self.lora_dropout(x)
+                            @ self.lora_A.transpose(0, 1)
+                            @ self.lora_B.transpose(0, 1)
+                        ) * self.scaling
+                    return result
+                else:
+                    return result
+
+        instance = QLoraLinear_cls.__new__(
+            QLoraLinear_cls
+        )  # Create a new instance of QLoraLinear_cls
+        instance.__init__(
+            *args, **kwargs
+        )  # Invoke the __init__ method of QLoraLinear_cls
+
+        """
+        # Check if QLoraLinear has a custom __init__ method
+        if hasattr(cls, "__init__"):
+            kwargs.pop("r", None)
+            kwargs.pop("quant_type", None)
+            # Invoke the __init__ method of QLoraLinear
+            cls.__init__(instance, *args, **kwargs)
+        """
+        return instance
+
+
+class QLoraLinear(metaclass=QLinear):
+    # LoRA implemented in a dense layer
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        **kwargs,
+    ):
+        pass
+
+
+class QQLoraLinear(nn.Module, LoRALayer):
```

Comment:
I find this one a bit more readable than the QLoraLinear one, but that may only be me.

Reply:
We need a class that inherit from 1) the proper module (either nn.Linear or the bnb.nn.LinearXbit) and 2) the LoRALayer class. When we do it with the solution QQLoraLinear we have to use nn.Module and create an attribute self.linear that will instantiate the right module from 1). Hence it creates another level "self.linear" in the model will will put the weights in linear_key.linear.weight (same for key and value) which then 1) make things less clear 2) requires a remap of weights keys wrt the state_dict.
Solution QLoraLinear seems more complicated but it is not. The only trick is that we need a wrapper because when we use a metaclass Python requires the keyword "metaclass= " in second position otherwise we could have used QLoraLinear(metaclass=QLinear, LoRALayer) which is forbidden.
Anyway the code logic in the metaclass QLinear which embeds QLoraLinear_cls is the same, but in the end the model will carry a QLoraLinear_cls which has directly the weights in its linear_key, linear_value, linear_query.

Reply:
All good for me then, and if people don't understand they'll come and read these comments I guess :)